File: AutoLayoutVisualFormat/UIView+SWAutoLayout.m
//  Gihhub: https://github.com/SolaWing/AutoLayoutVisualFormatLanguage
//
//  UIView+SWAutoLayout.m
//  AutoLayoutVisualFormat
//
//  Created by SolaWing on 15/9/24.
//  Copyright © 2015年 SW. All rights reserved.
//

#import "UIView+SWAutoLayout.h"
#import "NSArray+SWAutoLayout.h"
#import "AutoLayoutFormatAnalyzer.h"

static NSUInteger getMaxIndexValueInFormat(NSString* format) {
    const char* it = format.UTF8String;
    unsigned long maxIndex = 0;
    unsigned long n;
    while ((it = strchr(it, '$'))) {
        ++it;
        n = strtoul(it, (char**)&it, 10);
        if (n > maxIndex) maxIndex = n;
    }
    return maxIndex;
}

@implementation UIView (SWAutoLayout)

- (NSArray<NSLayoutConstraint*>*)VFLConstraints:(NSString*)format andVarArg:(va_list*)ap {
    NSMutableArray* env = [NSMutableArray arrayWithObject:self];
    NSUInteger count = getMaxIndexValueInFormat(format) + 1;
    NSUInteger i = 1;
    while ( i++<count ){
        [env addObject: va_arg(*ap, id)];
    }
    return VFLViewConstraints(format, self, env);
}

- (NSArray<NSLayoutConstraint*>*)VFLConstraints:(NSString*)format, ... {
    va_list ap; va_start(ap, format);
    NSArray* constraints = [self VFLConstraints:format andVarArg:&ap];
    va_end(ap);

    return constraints;
}

- (NSArray<NSLayoutConstraint*>*)VFLInstall:(NSString*)format, ... {
    va_list ap; va_start(ap, format);
    NSArray* constraints = [self VFLConstraints:format andVarArg:&ap];
    va_end(ap);
    [constraints activateConstraints];
    return constraints;
}

- (NSArray<NSLayoutConstraint*>*)VFLFullInstall:(NSString*)format, ... {
    va_list ap; va_start(ap, format);
    self.translatesAutoresizingMaskIntoConstraints = NO;
    NSArray* constraints = [self VFLConstraints:format andVarArg:&ap];
    va_end(ap);
    [constraints activateConstraints];
    return constraints;
}

@end

File: VFL/UIView+SWAutoLayout.swift
//  Gihhub: https://github.com/SolaWing/AutoLayoutVisualFormatLanguage
//
//  UIView+SWAutoLayout.swift
//  AutoLayoutVisualFormat
//
//  Created by SolaWing on 15/12/13.
//  Copyright © 2015年 SW. All rights reserved.
//

import Foundation
import UIKit
#if canImport(AutoLayoutVisualFormat)
import AutoLayoutVisualFormat
#endif

public extension UIView {
    // first type is String conflict with interpolation. so add prefix label
    @nonobjc
    func VFLConstraints(format:String, args:[Any]) -> [NSLayoutConstraint] {
        var argsIncludeSelf : [Any] = [self]
        argsIncludeSelf.append(contentsOf: args)
        return VFLViewConstraints(format, self, argsIncludeSelf)
    }

    @nonobjc
    func VFLConstraints(format:String, _ args:Any...) -> [NSLayoutConstraint] {
        return self.VFLConstraints(format: format, args: args)
    }

    @nonobjc
    @discardableResult
    func VFLInstall(format:String, _ args:Any...) -> [NSLayoutConstraint] {
        let constraints = self.VFLConstraints(format: format, args: args)
        NSLayoutConstraint.activate(constraints)
        return constraints
    }

    @nonobjc
    @discardableResult
    func VFLFullInstall(format:String, _ args:Any...) -> [NSLayoutConstraint] {
        self.translatesAutoresizingMaskIntoConstraints = false
        let constraints = self.VFLConstraints(format: format, args: args)
        NSLayoutConstraint.activate(constraints)
        return constraints
    }

    // MARK: - interpolation
    func VFLConstraints(_ interpolation:VFLInterpolation) -> [NSLayoutConstraint] {
        let format = NSMutableString()
        let env = NSMutableArray(object: self)
        interpolation.resultInto(format, env)
        return VFLViewConstraints(format as String, self, env)
    }

    @discardableResult
    func VFLInstall(_ interpolation:VFLInterpolation) -> [NSLayoutConstraint] {
        let constraints = self.VFLConstraints(interpolation)
        NSLayoutConstraint.activate(constraints)
        return constraints
    }

    @discardableResult
    func VFLFullInstall(_ interpolation:VFLInterpolation) -> [NSLayoutConstraint] {
        let constraints = self.VFLConstraints(interpolation)
        self.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate(constraints)
        return constraints
    }
}

@available(iOS 9.0, *)
public extension UILayoutGuide {
    // first type is String conflict with interpolation. so add prefix label
    @nonobjc
    func VFLConstraints(format:String, args:[Any]) -> [NSLayoutConstraint] {
        var argsIncludeSelf : [Any] = [self]
        argsIncludeSelf.append(contentsOf: args)
        return VFLViewConstraints(format, self, argsIncludeSelf)
    }

    @nonobjc
    func VFLConstraints(format:String, _ args:Any...) -> [NSLayoutConstraint] {
        return self.VFLConstraints(format: format, args: args)
    }

    @nonobjc
    @discardableResult
    func VFLInstall(format:String, _ args:Any...) -> [NSLayoutConstraint] {
        let constraints = self.VFLConstraints(format: format, args: args)
        NSLayoutConstraint.activate(constraints)
        return constraints
    }

    @nonobjc
    @discardableResult
    func VFLFullInstall(format:String, _ args:Any...) -> [NSLayoutConstraint] {
        let constraints = self.VFLConstraints(format: format, args: args)
        NSLayoutConstraint.activate(constraints)
        return constraints
    }

    // MARK: - interpolation
    func VFLConstraints(_ interpolation:VFLInterpolation) -> [NSLayoutConstraint] {
        let format = NSMutableString()
        let env = NSMutableArray(object: self)
        interpolation.resultInto(format, env)
        return VFLViewConstraints(format as String, self, env)
    }

    @discardableResult
    func VFLInstall(_ interpolation:VFLInterpolation) -> [NSLayoutConstraint] {
        let constraints = self.VFLConstraints(interpolation)
        NSLayoutConstraint.activate(constraints)
        return constraints
    }

    @discardableResult
    func VFLFullInstall(_ interpolation:VFLInterpolation) -> [NSLayoutConstraint] {
        let constraints = self.VFLConstraints(interpolation)
        NSLayoutConstraint.activate(constraints)
        return constraints
    }
}
